src/stac_utils/bsd.py

```diff
@@ -0,0 +1,150 @@
+import hmac, hashlib
+import os, time
+import sys
+
+import xmltodict
+import requests
+import pandas as pd
+
+from insert_leads import insert_hot_lead
+from datetime import datetime, date, timedelta
+
+if "BSD_API_SECRET" not in os.environ:
+    # Try to load from .env
+    from dotenv import load_dotenv
+
+    load_dotenv()
+
+try:
+    api_secret = os.environ['BSD_API_SECRET']
+    BSD_API_ID = os.environ['BSD_API_ID']
+    BSD_URL = os.environ['BSD_URL']
+except KeyError:
+    sys.exit('Cannot find BSD API Secret')
+
+api_ver = 2  # BSD API ID - always 2
+signup_form_id = "115"  # NC Volunteer Form ID from BSD
+
+
+# Generating the API MAC necessary for the BSD API
+def generate_api_mac(time, url, params_str=None):
+    root_params = 'api_ver=2&api_id=' + BSD_API_ID \
+                  + '&api_ts=' + time
+
+    params = root_params
+
+    if params_str != None:
+        params += "&" + params_str
+
+    signing_str = BSD_API_ID + os.linesep \
+                  + time + os.linesep \
+                  + url + os.linesep \
+                  + params
+
+    api_mac = hmac.new(api_secret.encode(), signing_str.encode(), hashlib.sha1).hexdigest()
+
+    print(signing_str)
+    print(api_mac)
+
+    return api_mac
+
+
+# Get volunteer signups from the NC Volunteer Form from the previous day.
+# Returns a pandas DataFrame containing sign in data.
+def get_volunteer_signups():
+    current_time = str(int(time.time()))
+
+    url = '/page/api/signup/get_signups_by_form_id'
+    params_str = "signup_form_id=" + signup_form_id
+
+    columns = ['ID', 'First Name', 'Last Name', 'Phone', 'Email', 'Zip', 'VR', 'Canvass', 'Calls', 'Texts',
+               'Attend Events', 'Donate', 'Signup Date']
+    api_mac = generate_api_mac(current_time, url, params_str)
+
+    params = {
+        'api_ver': api_ver,
+        'api_id': BSD_API_ID,
+        'api_ts': current_time,
+        'api_mac': api_mac,
+        'signup_form_id': signup_form_id
+    }
+
+    today = datetime.now().date()
+
+    resp = requests.get(get_full_url(url), params=params)
+
+    if resp.status_code == requests.codes.ok:
+        signup_data = []
+        print(resp.text)
+
+        # Response from BSD is an XML structure so we need to parse it
+        xml = xmltodict.parse(resp.text, xml_attribs=False)
```

Comment:
in my opinion you should do this for your transform_response, similar to what's in the actionnetwork class, and then just spit the resulting data dict out and call it a day, and not include anything after this point at all